python/cudf_polars/cudf_polars/containers/dataframe.py

```diff
@@ -0,0 +1,173 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""A dataframe, with some properties."""
+
+from __future__ import annotations
+
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+import polars as pl
+
+import cudf._lib.pylibcudf as plc
+
+from cudf_polars.containers.column import Column
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping, Sequence, Set
+
+    from typing_extensions import Self
+
+    import cudf
+
+    from cudf_polars.containers.scalar import Scalar
+
+
+__all__: list[str] = ["DataFrame"]
+
+
+class DataFrame:
+    """A representation of a dataframe."""
+
+    columns: list[Column]
+    scalars: list[Scalar]
+    table: plc.Table | None
+
+    def __init__(self, columns: Sequence[Column], scalars: Sequence[Scalar]) -> None:
+        self.columns = list(columns)
+        self._column_map = {c.name: c for c in self.columns}
+        self.scalars = list(scalars)
+        if len(scalars) == 0:
+            self.table = plc.Table([c.obj for c in columns])
+        else:
+            self.table = None
+
+    __iter__ = None
```

Comment:
FWIW I think this is redundant if __getitem__ isn't implemented, but no harm in making this explicit.